storage/remote: Release two goroutines from endless loops

lzhfromustc opened the pull request:
***What this PR does***
Change the structure and methods of `storage/remote.Storage` and `storage/remote.WriteStorage`, to release two goroutines from endless loops.
This PR should have no side effects.

***Problem with Storage***
In the following function, a new `*Storage` is created, with a `*logging.Deduper`:
https://github.com/prometheus/prometheus/blob/eda5f8ca5f2223620cae36627919bab58c78634f/storage/remote/storage.go#L64-L76
The problem is that `logging.Dedupe()` creates a goroutine running `logging.Deduper.run()`, and it won't end until `logging.Deduper.Stop()` is called.
However, after we create the `logging.Deduper` in `NewStorage()`, we never call `logging.Deduper.Stop()`

**How is the problem fixed**
Add a new field `loggerStop` to `*=Storage`, which stores the function `logging.Deduper.Stop()`. This function will be called in `Storage.Close()`

***Problem with WriteStorage***
When a new `WriteStorage` is created, a new goroutine will start and run `WriteStorage.run()`
https://github.com/prometheus/prometheus/blob/eda5f8ca5f2223620cae36627919bab58c78634f/storage/remote/write.go#L96-L106
The problem is that `WriteStorage.run()` never ends, because `time.Ticker.Stop()` is never reached.

**How is the problem fixed**
Add a new field `ticker` to `WriteStorage`, and store the `time.Ticker` in it. In `WriteStorage.Close()`, the ticker will be stopped.

roidelapluie reviewed (commented):
Thanks. I'll let our remote storage maintainers have a closer look.

cstyan commented on storage/remote/storage.go:
```diff
@@ -181,6 +184,7 @@ func (s *Storage) Appender(ctx context.Context) storage.Appender {
 
 // Close the background processing of the storage queues.
 func (s *Storage) Close() error {
+	s.loggerStop()
```
we can just call `s.logger.Stop()` here

lzhfromustc replied:
Sorry again for my late response!
In this package, I can't call `s.logger.Stop()` because `s.logger` is `log.Logger`, and `log.Logger` is an interface with only one method `Log()` (code is below).
```go
// .../pkg/mod/github.com/go-kit/log@v0.1.0/log.go:10
type Logger interface {
	Log(keyvals ...interface{}) error
}
```

cstyan replied:
In that case we can just change the field on the storage struct to be `logger *logging.Deduper`

lzhfromustc replied:
Got it. I pushed the change using type `*logging.Deduper` in struct definition and using `s.logger.Stop()`

cstyan commented:
It looks like there's some `gofmt` failures, are you using go 1.16?

lzhfromustc commented:
Thanks. I gofmted it using 1.16. Now it should be ok

cstyan reviewed (commented):
Try running the tests locally with the `-race` flag, `TestNoDuplicateReadConfigs` is failing due to a race condition with the way you've implemented the ticker lifecycle. 

Have a look at how we manage the tickers and other goroutines in `queue_manager.go`. We include an additional channel on the `QueueManager` struct `quit`, and when we start extra routines from the queue manager the functions select on their tickers as well as that quit channel. This way closing the over arching struct can close that channel, and the other routines can exit properly as well.

lzhfromustc commented:
I see, there is a data race problem I introduced.
Thanks for the `quit` example in `QueueManager`! I fixed my commit by adding the field `quit` in `WriteStorage`. 
Now the goroutines create their own ticker (to avoid data race). They can exit and stop the ticker after `quit` is closed in `WriteStorage.Close()`

cstyan commented on storage/remote/write.go:
```diff
@@ -100,8 +102,12 @@ func NewWriteStorage(logger log.Logger, reg prometheus.Registerer, walDir string
+		select {
+		case <-ticker.C:
+			rws.samplesIn.tick()
+		case <-rws.quit:
```
you still need to break the outer loop and/or return here
